doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,426 @@
+```c
+__owur int ossl_quic_packet_sent(SSL_QUIC_TX_PACKETISER *tx,
+                                 OSSL_QUIC_PACKET *packet);
+__owur int ossl_quic_packet_acked(SSL_QUIC_TX_PACKETISER *tx,
+                                  OSSL_QUIC_PACKET *packet);
+__owur int ossl_quic_packet_lost(SSL_QUIC_TX_PACKETISER *tx,
+                                 OSSL_QUIC_PACKET *packet);
+__owur int ossl_quic_packet_discarded(SSL_QUIC_TX_PACKETISER *tx,
+                                      OSSL_QUIC_PACKET *packet);
+```
+
+#### Notes
+
+| Name here | Name in ACK Manager |
+| --- | --- |
+| `ossl_quic_packet_sent` | `QUIC_ACKM_on_tx_ack_packet` |
+| `ossl_quic_packet_acked` | `on_acked` |
+| `ossl_quic_packet_lost` | `on_lost` |
+| `ossl_quic_packet_discarded` | `on_discarded` |
+
+Packets
+-------
+
+Packets formats are defined in [RFC 9000 17.1 Packet Formats].
+
+### Packet types
+
+QUIC supports a number of different packets.  The combination of packets of
+different types as per [RFC 9000 12.2 Coalescing Packets], is done by the
+record layer.
+
+#### Version Negotiation Packet
+
+Refer to [RFC 9000 17.2.1 Version Negotiation Packet].
+
+#### Initial Packet
+
+Refer to [RFC 9000 17.2.2 Initial Packet].
+
+#### Handshake Packet
+
+Refer to [RFC 9000 17.2.4 Handshake Packet].
+
+#### App Data 0-RTT Packet
+
+Refer to [RFC 9000 17.2.3 0-RTT].
+
+#### App Data 1-RTT Packet
+
+Refer to [RFC 9000 17.3.1 1-RTT].
+
+#### Retry Packet
+
+Refer to [RFC 9000 17.2.5 Retry Packet.
+
+Packetisiation and Processing
+-----------------------------
+
+### Application data frames
+
+The packetiser builds application data frames after requesting a specific
+amount of application data.  If insufficient data is available, or buffer
+boundaries prevent fulfilling the entire request, the stream send buffer module
+is free to return a smaller amount of data.
+
+```c
+int ossl_quic_ssb_request_data(SSL_CONNECTION *conn, size_t ideal_length,
+                               const void **data, size_t *actual_length);
+```
+
+### Retransmission
+
+When a packet is determined to be lost by the ACK Manager, the
+`ossl_quic_packet_lost()` function will be called.  This function will
+extract the frame references from the packet and re-queue them for
+transmission as if `ossl_quic_packetise()` had been called for each
+frame followed by `ossl_quic_send_packets()`.  Frames that need to be
+retransmitted will be be considered higher priority than other pending
+frames, although both types are available to construct packets from.
+Moreover, any such constructed packets will not be subject to a delay
+before transmission.
+
+### Restricting packet sizes
+
+Three factors impact the size of packets that can be sent:
+
+* MTU restricting packet sizes
+* Flow control
+* Congestion control
+
+The MTU limits the size of an individual packet, the other two limit the
+total amount of data that can be sent.  The packetiser needs to query the
+current limits using the `ossl_quic_flow_maximum_size()`,
```

Comment:
I am unsure whether this isn't too late - IMO the flow control limits should be limited already when writing to the stream send buffers. I.e, on the SSL_ API front end call.